cmd/agent/main.go

```diff
@@ -24,54 +27,64 @@ func main() {
 		os.Exit(1)
 	}
 
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var wg sync.WaitGroup
+
 	metricsStore := storage.NewMetricsCollection()
 	metricsRepo := repositories.NewAgentRepository(metricsStore, agentCfg.Log)
-	metricsService := agent.NewAgentService(metricsRepo, agentCfg.Log)
+	metricsService := agent.NewAgentService(metricsRepo, agentCfg.Log, &agentCfg.Key)
+
+	workerPool := worker.NewWorkerPool(agentCfg.RateLimit)
 
 	pollTicker := time.NewTicker(agentCfg.PollInterval)
 	defer pollTicker.Stop()
 	reportTicker := time.NewTicker(agentCfg.ReportInterval)
 	defer reportTicker.Stop()
 
-	originalReportInterval := agentCfg.ReportInterval
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		<-sigChan
+		cancel()
+	}()
 
-	for {
-		select {
-		case <-pollTicker.C:
-			if err = metricsService.MetricsService.Collect(); err != nil {
-				agentCfg.Log.ErrorContext(context.Background(),
-					"failed to collect the metrics: ",
-					helpers.ErrAttr(err))
-			}
-		case <-reportTicker.C:
-			retries := 3
-			backoffIntervals := []time.Duration{1 * time.Second, 3 * time.Second, 5 * time.Second}
-			success := false
-
-			for attempt := range backoffIntervals {
-				err = metricsService.MetricsService.Send(agentCfg.ServerAddr)
-				if err == nil {
-					metricsStore.Clear()
-					reportTicker.Reset(originalReportInterval)
-					success = true
-					break
+	go func() {
```

Comment:
To implement graceful shutdown properly, you should make sure this goroutine is finished before you stop the process

Reply:
Yeah, makes sense, updated the code and tracked the units of work with a wait group. Thank you!